## Part 2 - Data Transformation & Cleaning

#### Import the libraries

In [1]:
import pandas as pd
import numpy as np
import os
import joblib #serialize the dataframes produced from code execution
import tqdm.notebook as tqdm #monitor the progress of python for loops and pandas.apply() method
import re #python regex module

c:\users\spano\miniconda3\lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
c:\users\spano\miniconda3\lib\site-packages\numpy\.libs\libopenblas.JPIJNSWNNAN3CE6LLI5FWSPHUT2VXMTH.gfortran-win_amd64.dll
c:\users\spano\miniconda3\lib\site-packages\numpy\.libs\libopenblas.TXA6YQSD3GCQQC22GEQ54J2UDCXDXHWN.gfortran-win_amd64.dll
  warnings.warn("loaded more than 1 DLL from .libs:"


#### Import the dataset

In [5]:
"""
dataset_part_1_25012020=final_dataset_49393_movies_25012020 are the same table files produced from Part 1.
"""
dataset=pd.read_pickle(os.path.join(os.getcwd(), "dataset_part_1_25012020.pkl"))
dataset=dataset.reset_index(drop=True)
dataset.shape

(49393, 10)

#### -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --

### Section 2.1: Cleaning columns and creating new ones based on research needs
To follow through this section you are advised to read the comments on top of each code block

In [6]:
"""2.1.1: Create a separate column for the year the movie was published"""
pattern = '\((\d{4})\)'
dataset['year'] =dataset.title.str.extract(pattern, expand=False) #False returns a series
dataset['year'].iloc[0]

'1995'

In [7]:
"""2.1.2: Clean the year (i.e 2001) from each title"""
exp = r'\(\d\d\d\d.'
dataset['title'] = dataset['title'].apply(lambda x: re.sub(exp,"",x).strip())
dataset.head()

title                            genres  rating  \
0                    Toy Story  [Adventure, Animation, Children]    3.89   
1                      Jumanji    [Adventure, Children, Fantasy]    3.25   
2             Grumpier Old Men                 [Comedy, Romance]    3.17   
3            Waiting to Exhale          [Comedy, Drama, Romance]    2.87   
4  Father of the Bride Part II                          [Comedy]    3.08   

                               imdb_url  \
0  http://www.imdb.com/title/tt0114709/   
1  http://www.imdb.com/title/tt0113497/   
2  http://www.imdb.com/title/tt0113228/   
3  http://www.imdb.com/title/tt0114885/   
4  http://www.imdb.com/title/tt0113041/   

                                         reviews_url  \
0  http://www.imdb.com/title/tt0114709/reviews?sp...   
1  http://www.imdb.com/title/tt0113497/reviews?sp...   
2  http://www.imdb.com/title/tt0113228/reviews?sp...   
3  http://www.imdb.com/title/tt0114885/reviews?sp...   
4  http://www.imdb.com/title/tt0113041/reviews?sp...   

                                              actors  \
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...   
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst,...   
2  [Walter Matthau, Jack Lemmon, Sophia Loren, An...   
3  [Whitney Houston, Angela Bassett, Loretta Devi...   
4  [Steve Martin, Diane Keaton, Martin Short, Kim...   

                                                plot imdb_rating  \
0  \n                    A cowboy doll is profoun...         8.3   
1  \n                    When two kids find and p...         7.0   
2  \n                    John and Max resolve to ...         6.7   
3  \n                    Based on Terry McMillan'...         5.9   
4  \n                    George Banks must deal n...         6.0   

          director                                            reviews  year  
0    John Lasseter  [Andy's toys live a reasonable life of fun and...  1995  
1     Joe Johnston  [The smash hit 1995 film Jumanji- based on the...  1995  
2    Howard Deutch  [Sequels are rarely half as good as the origin...  1995  
3  Forest Whitaker  [I read some of the other comments comcerning ...  1995  
4    Charles Shyer  [I enjoyed this film, as I did Father of the B...  1995

In [8]:
"""
2.1.3: Replace (,The) & (,A), (, An) from the end of the string, at the front + Remove any text inside paranthesis
As we can see "Dark Knight Rises, The" (index 17974) is not correct and should be corrected to "The Dark Knight Rises
"""
dataset[dataset.director=="Christopher Nolan"]

title                      genres  rating  \
2408                Following  [Crime, Mystery, Thriller]    3.77   
4038                  Memento         [Mystery, Thriller]    4.15   
5187                 Insomnia      [Action, Crime, Drama]    3.48   
10062           Batman Begins       [Action, Crime, IMAX]    3.93   
11291           Prestige, The    [Drama, Mystery, Sci-Fi]    4.09   
12422        Dark Knight, The      [Action, Crime, Drama]    4.17   
15361               Inception      [Action, Crime, Drama]    4.16   
17974  Dark Knight Rises, The  [Action, Adventure, Crime]    3.98   
22130            Interstellar              [Sci-Fi, IMAX]    4.09   
25087               Doodlebug          [Fantasy, Mystery]    3.16   
41024                 Dunkirk   [Action, Drama, Thriller]    3.80   

                                   imdb_url  \
2408   http://www.imdb.com/title/tt0154506/   
4038   http://www.imdb.com/title/tt0209144/   
5187   http://www.imdb.com/title/tt0278504/   
10062  http://www.imdb.com/title/tt0372784/   
11291  http://www.imdb.com/title/tt0482571/   
12422  http://www.imdb.com/title/tt0468569/   
15361  http://www.imdb.com/title/tt1375666/   
17974  http://www.imdb.com/title/tt1345836/   
22130  http://www.imdb.com/title/tt0816692/   
25087  http://www.imdb.com/title/tt0411302/   
41024  http://www.imdb.com/title/tt5013056/   

                                             reviews_url  \
2408   http://www.imdb.com/title/tt0154506/reviews?sp...   
4038   http://www.imdb.com/title/tt0209144/reviews?sp...   
5187   http://www.imdb.com/title/tt0278504/reviews?sp...   
10062  http://www.imdb.com/title/tt0372784/reviews?sp...   
11291  http://www.imdb.com/title/tt0482571/reviews?sp...   
12422  http://www.imdb.com/title/tt0468569/reviews?sp...   
15361  http://www.imdb.com/title/tt1375666/reviews?sp...   
17974  http://www.imdb.com/title/tt1345836/reviews?sp...   
22130  http://www.imdb.com/title/tt0816692/reviews?sp...   
25087  http://www.imdb.com/title/tt0411302/reviews?sp...   
41024  http://www.imdb.com/title/tt5013056/reviews?sp...   

                                                  actors  \
2408   [Jeremy Theobald, Alex Haw, Lucy Russell, John...   
4038   [Guy Pearce, Carrie-Anne Moss, Joe Pantoliano,...   
5187   [Al Pacino, Martin Donovan, Oliver 'Ole' Zemen...   
10062  [Christian Bale, Michael Caine, Liam Neeson, K...   
11291  [Hugh Jackman, Christian Bale, Michael Caine, ...   
12422  [Christian Bale, Heath Ledger, Aaron Eckhart, ...   
15361  [Leonardo DiCaprio, Joseph Gordon-Levitt, Elle...   
17974  [Christian Bale, Gary Oldman, Tom Hardy, Josep...   
22130  [Ellen Burstyn, Matthew McConaughey, Mackenzie...   
25087                                  [Jeremy Theobald]   
41024  [Fionn Whitehead, Damien Bonnard, Aneurin Barn...   

                                                    plot imdb_rating  \
2408   \n                    A young writer who follo...         7.5   
4038   \n                    A man with short-term me...         8.4   
5187   \n                    Two Los Angeles homicide...         7.2   
10062  \n                    After training with his ...         8.2   
11291  \n                    After a tragic accident,...         8.5   
12422  \n                    When the menace known as...         9.0   
15361  \n                    A thief who steals corpo...         8.8   
17974  \n                    Eight years after the Jo...         8.4   
22130  \n                    A team of explorers trav...         8.6   
25087  \n                    In his squalid apartment...         7.1   
41024  \n                    Allied soldiers from Bel...         7.9   

                director                                            reviews  \
2408   Christopher Nolan  [The debut that plucked from obscurity one of ...   
4038   Christopher Nolan  [Thank Goodness I didn't read the reviews post...   
5187   Christopher Nolan  [I didn't know that 'Insomnia' was based on a ...   
10062  

In [9]:
"""Continue of step 2.1.3 to apply the regex transformation of the movie title"""
dataset['title']=dataset['title'].str.replace(r'(.*?),?\s*(The|A|An|Les)?(?=\s*\(.*\)\s*|$).*', r'\2 \1')
dataset['title']=dataset['title'].str.strip()

<ipython-input-9-64f41b4541cf>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  dataset['title']=dataset['title'].str.replace(r'(.*?),?\s*(The|A|An|Les)?(?=\s*\(.*\)\s*|$).*', r'\2 \1')


In [11]:
"""
(1) title: Dark Knight, The (index 12422), has been successfully changed to The Dark Knight
(2) title: Dark Knight Rises, The (index 17974), has been successfully changed to The Dark Knight Rises
"""
print(dataset["title"].iloc[12422])
print(dataset["title"].iloc[17974])

The Dark Knight
The Dark Knight Rises


In [12]:
"""
2.1.4: Clean some punctuation mistakes in columns plot summary and user reviews.
This code block will be refactored in later versions of the notebook.
"""
dataset['plot']=dataset['plot'].apply(lambda x: re.sub(" +", " ", x.strip().replace(',', ', ').replace('.', '. ').replace('?', '? ').replace('!', '! ').replace('\n                    See full summary\xa0»', '').rstrip()))
dataset['reviews']=dataset['reviews'].apply(lambda x: [re.sub(" +", " ", user_review.strip().replace(',', ', ').replace('.', '. ').replace('?', '? ').replace('!', '! ').rstrip()) for user_review in x])

In [14]:
"""After cleaning the plot of the movies using the above code block, check the plot of a random movie."""
pd.set_option('display.max_colwidth', None)
dataset['plot'][dataset['title']=="Come Look at Me"]

33181    Before New Year Eve the windows of vast city are glowing from within by colored Christmas lights. It seems that behind each of these windows carefully stored favorite holiday. It seems that . . .
Name: plot, dtype: object

In [15]:
"""
2.1.5: Remove empty summaries. Movies with empty plot summaries should be removed.
In total 176 movies should be removed, due to UNKNOWN plot summary text.
"""
print(f"The number of movies with an unregistered/empty plot from IMDB: {len(dataset[dataset['plot']=='Add a Plot »'])}")
dataset=dataset[dataset['plot']!='Add a Plot »']
dataset.shape

The number of movies with an unregistered/empty plot from IMDB: 176


(49217, 11)

In [16]:
"""2.1.6: Create the positive/negative column"""
dataset.loc[:, 'sentiment_value']=dataset['rating']
dataset.loc[:, 'sentiment_value'].loc[dataset.rating >=2.5] = 1
dataset.loc[:, 'sentiment_value'].loc[dataset.rating <2.5] = 0

c:\users\spano\miniconda3\lib\site-packages\pandas\core\indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)
c:\users\spano\miniconda3\lib\site-packages\pandas\core\indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


In [17]:
"""2.1.7 Create the column Combined Features"""
def combine_features(row):
    """
    Purpose: This function will create the column 'movie_features'. The column will combine text content from multiple columns of the dataframe.
             Specifically the text combined derives from the following columns:
             1) Title,
             2) Actors,
             3) Director,
             4) Plot summary,
             5) Genres
    Argument: Dataset row
    Output: The unified text of five columns into one column.
    """
    return row['title'] + " " + ' '.join(map(str, row['actors'])) + " " + row['director'] + " " + row['plot'] + " " + ' '.join(row['genres'])

dataset["movie_features"]=dataset.apply(combine_features, axis=1)
dataset["movie_features"].iloc[0]

"Toy Story Tom Hanks Tim Allen Don Rickles Jim Varney Wallace Shawn John Ratzenberger Annie Potts John Morris Erik von Detten Laurie Metcalf R. Lee Ermey Sarah Freeman Penn Jillette Jack Angel Spencer Aste John Lasseter A cowboy doll is profoundly threatened and jealous when a new spaceman figure supplants him as top toy in a boy's room. Adventure Animation Children"

#### -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --

### Section 2.2: Remove movies that belong to low-frequency (sparse) genre tags & Replace plots and movie_features with new content for some movies.
To follow through this section you are advised to read the comments on top of each code block

In [18]:
"""
Execute the code below to get an overview of the genres distribution. This step is already applied to the dataset,
so re-executing the code block will not affect the rows of the dataset.

Having cut the most scarse occurences of genres it is still obvious that genres "Drama" & "Comedy" belong to 40% of the movies.
A good approach is either to up-sample the dataset or down-sample it.
What we chose was to down-sample the two dominant genres "Drama" & "Comedy". However, in the sub-part 3.2.
We used the imbalanced dataset to train and test the keras text classification models.
"""
print(round(dataset['genres'].explode().value_counts(normalize=True) * 100,3))
genre_frequency=round(dataset['genres'].explode().value_counts(normalize=True) * 100,3)
genre_tags_remove=genre_frequency[genre_frequency<1.00].index.tolist() #remove the genres with less than 1% of frequency
print(f"genre tags to remove: {genre_tags_remove}")

# STEP 1: Remove genres less than 1% frequency
dataset['reduced_genres']=dataset['genres'].apply(lambda row:[val for val in row if val not in genre_tags_remove])

# STEP 2: Find indexes with EMPTY LISTS
dataset_empty_lists=dataset[dataset['reduced_genres'].apply(lambda c: c==[])]
remove_indices=dataset_empty_lists.index.to_list()
print(f"indexes to remove: {remove_indices} (total: {len(remove_indices)})")

# STEP 3: Remove the indexes with EMPTY LISTS
dataset_frequent_genres=dataset[~dataset.index.isin(remove_indices)]
dataset_frequent_genres=dataset_frequent_genres.reset_index(drop=True)
print(f"The shape of the dataset after removing the low-frequency genre tags: {dataset_frequent_genres.shape}")
round(dataset_frequent_genres['reduced_genres'].explode().value_counts(normalize=True) * 100,3)

Drama          24.982
Comedy         15.684
Action          7.478
Romance         7.105
Thriller        6.925
Horror          5.665
Crime           5.356
Documentary     4.300
Adventure       4.223
Sci-Fi          2.978
Mystery         2.703
Children        2.490
Animation       2.354
Fantasy         2.266
War             1.680
Western         1.353
Musical         1.151
Film-Noir       0.388
Short           0.264
Family          0.151
Biography       0.134
Music           0.127
History         0.125
IMAX            0.060
Sport           0.044
Adult           0.005
News            0.003
Reality-TV      0.003
Name: genres, dtype: float64
genre tags to remove: ['Film-Noir', 'Short', 'Family', 'Biography', 'Music', 'History', 'IMAX', 'Sport', 'Adult', 'News', 'Reality-TV']
indexes to remove: [717, 1081, 1901, 3113, 3199, 4237, 4969, 5591, 7112, 14256, 18022, 19599, 22408, 46551, 46555, 46559, 46560, 46561, 46562, 46575, 46579, 46580, 46584, 46586, 46587, 46656, 46746, 46788, 46819, 46836,

Drama          25.313
Comedy         15.892
Action          7.577
Romance         7.199
Thriller        7.016
Horror          5.740
Crime           5.427
Documentary     4.357
Adventure       4.279
Sci-Fi          3.018
Mystery         2.739
Children        2.523
Animation       2.385
Fantasy         2.296
War             1.702
Western         1.371
Musical         1.166
Name: reduced_genres, dtype: float64

In [29]:
"""Apply some changes to plot summary and reviews of the dataset created in part 2 with the minority genres filtered out"""
# REPLACE THE TEXT OF THOSE MOVIE PLOTS & FEATURES WITH THOSE BELOW, OTHERWISE THE PREPROCESSING WON'T COMPLETE CORRECTLY
dataset_frequent_genres.iloc[10802,6]='A documentary looking at the Nigga word'
dataset_frequent_genres.iloc[15148,6]="A shy Arun works as an accountant. On daily basis he travels by a bus to office where he sees Prabha who works in the same premises. Arun starts to like Prabha and follows her frequently but hasn't got the confidence in him to approach her. When ever Arun gains confidence to share about his feelings to Prabha her friend Nagesh steps in spoiling Arun's plans. In order to win his love for Prabha, Arun travels all the way to Lonavala to take the help of expert Colonel Julius Naganderanath Wilfred Singh. During this phase Prabha to starts missing Arun."
dataset_frequent_genres.iloc[31994,6]="MARTIN ARMSTRONG, once a US based trillion dollar financial adviser, used the number pi to predict economic turning points with precision. When some big New York bankers asked him to join the club to help them to take over Russia, he refused to join the manipulation. A few days later the FBI stormed his offices accusing him of a 3 billion dollar Ponzi Scheme - an attempt to stop him talking about the real Ponzi Scheme of debts that the US has build up over the years and which he thinks starts to collapse after October 1, 2015, a major pi turning point he is predicting."
dataset_frequent_genres.iloc[32858,6]="Notorious gangster asks a career thief by the name of Nandu to steal a package worth over one crore rupees for a fee, which Nandu does so. Things do not go as planned, when Nandu finds out that the package is worth more than twice the amount, he refuses to hand over the package to Pinky, until he raises his fees. An enraged Pinky instructs his men to locate Nandu, find the package and bring it back, and thereafter kill Nandu. While on the run from Pinky's gangsters, the police find out that Nandu was involved in this theft, and they launch a man-hunt. What Nandu does not know that the package that he has on his person does not contain money or gold, but a nuclear bomb that is set to explode."
dataset_frequent_genres.iloc[33094,6]="Before New Year Eve the windows of vast city are glowing from within by colored Christmas lights. It seems that behind each of these windows carefully stored favorite holiday. It seems that each of them are waiting for the New Year's wish fulfillment. By the silhouette of an elderly woman at one of the windows of the old house in Moscow, neighbors in the yard probably accustomed as something unchangeable. Sophia Ivanovna ten years does not get up from her chair, all day looking out the window, glues paper figurines and listening to Dickens, who reads aloud to her, her only daughter Tanya. The only heiress, Tanya, whom Sophia Ivanovna must to pass all the family jewels, seems resigned to the position of an old maid, and her whole life consists only of the care of the sick mother."
dataset_frequent_genres.iloc[36101,6]="Mr. Giant has kidnapped the brilliant Dr. Van Kohler and is planning to use the Doctor's invention, the N-bomb, to hold the world hostage. The only one who can foil Mr. Giant's evil scheme is Agent 00, a three foot tall filipino martial arts master, expert marksman, top-class romancer and all-around superspy. Can Agent 00 rescue Dr. Kohler before it's too late?"
dataset_frequent_genres.iloc[36670,6]="HEADER portrays the grueling psychological journey taken by ATF Agent Stewart Cummings. On the surface, Stewart struggles to solve a string of bizarre murders, but in secret, his life falls into a world of corruption that's impossible to escape. Deceit, rape, and murder spiral out of control triggering a hellish conclusion that defies description."
dataset_frequent_genres.iloc[47726,6]="Raj is one of city's top lawyer known never to loose a case. He his introduced to Sargam during the launch of her new album by Ashwin Mehta owner of a music company. Next day Ashwin is found murdered and Tarang is held for the murder, Sargam and Tarang are childhood friends and she approaches Raj to fight for his case. During this period Raj notices that Tarang suffers from mental illness of split personality and when he suffers the attack he is not Tarang but Ranjeet and doesn't remember things he does when he turns Ranjeet. Tarang had killed Ashwin as the previous night he saw him forcing on Sargam but doesn't remember killing him as he was Ranjeet that time. Raj wins the case in court by proving Tarang innocent as he is suffering from mental illness. Tarang is due to release and gives Raj a shock to that he had planned everything regarding split personality as he was caught red handed on the site of murder."

dataset_frequent_genres.iloc[8102,12]="Spinning Boris Jeff Goldblum Anthony LaPaglia Liev Schreiber Boris Lee Krutonog Svetlana Efremova Shauna MacDonald Gregory Hlady Vladimir Radian Ilia Volok Konstantin Kazakov Judah Katz Maria Syrgiannis Ola Sturik Gillian Vanderburgh Serge Timokhin Roger Spottiswoode Early in 1996, three Republican campaign operatives take a job in secret assisting Boris Yeltsin's reelection. Once in Moscow, they find he's polling at 6 percent with the election a few months away. While Dick Dresner wants to go home, George Gorton and Joe Shumate vote to stay. First, they must get someone's attention; they succeed finally with Yeltsin's daughter. Then it's polling, focus groups, messages and spin. Even as Yeltsin's numbers go up, the trio are unsure who hired them and whether Yeltsin's allies have a different plan in mind than election victory. When the going gets toughest, it's Gorton who puts a spin on our stake: democracy and capitalism must win. Comedy Drama"
dataset_frequent_genres.iloc[10802,12]="The N Word F. Lee Bailey Sandra Bernhard Donald Bogle Todd Boyd Elaine Brown LeVar Burton George Carlin Morris Chestnut Chuck D Johnnie L. Cochran Jr. Stanley Crouch Damon Dash Dr. Dre Eazy-E Laurence Fishburne Todd Williams A documentary looking at the Nigga word Documentary"
dataset_frequent_genres.iloc[15148,12]="Chhoti Si Baat Ashok Kumar Vidya Sinha Amol Palekar Asrani Nandita Thakur Rajan Haksar Ashim Kumar Devendra Khandelwal Baba Majgoakar Noni Ganguly C.S. Dubey Amol Sen R.S. Chopra Sudarshan Sahni Milon Mukerji Basu Chatterjee A shy Arun works as an accountant. On daily basis he travels by a bus to office where he sees Prabha who works in the same premises. Arun starts to like Prabha and follows her frequently but hasn't got the confidence in him to approach her. When ever Arun gains confidence to share about his feelings to Prabha her friend Nagesh steps in spoiling Arun's plans. In order to win his love for Prabha, Arun travels all the way to Lonavala to take the help of expert Colonel Julius Naganderanath Wilfred Singh. During this phase Prabha to starts missing Arun. Comedy Romance"
dataset_frequent_genres.iloc[31994,12]="The Forecaster Martin Armstrong Vicky Armstrong Oliver Brown Michael Campbell Larry Edelson Tony Godin Nigel Kirwan Barclay Leib Neill MacPherson Thomas Sjoblom Marcus Vetter MARTIN ARMSTRONG, once a US based trillion dollar financial adviser, used the number pi to predict economic turning points with precision. When some big New York bankers asked him to join the club to help them to take over Russia, he refused to join the manipulation. A few days later the FBI stormed his offices accusing him of a 3 billion dollar Ponzi Scheme - an attempt to stop him talking about the real Ponzi Scheme of debts that the US has build up over the years and which he thinks starts to collapse after October 1, 2015, a major pi turning point he is predicting. Documentary"
dataset_frequent_genres.iloc[32858,12]="Daud Sanjay Dutt Urmila Matondkar Paresh Rawal Neeraj Vora Ram Mohan Ashish Vidyarthi Manoj Bajpayee Rajeev Mehta Jeetendra Shastri Sumukhi Tarzan Vineeth Narsing Yadav Rana Jung Bahadur Sunil Shende Ram Gopal Varma Notorious gangster asks a career thief by the name of Nandu to steal a package worth over one crore rupees for a fee, which Nandu does so. Things do not go as planned, when Nandu finds out that the package is worth more than twice the amount, he refuses to hand over the package to Pinky, until he raises his fees. An enraged Pinky instructs his men to locate Nandu, find the package and bring it back, and thereafter kill Nandu. While on the run from Pinky's gangsters, the police find out that Nandu was involved in this theft, and they launch a man-hunt. What Nandu does not know that the package that he has on his person does not contain money or gold, but a nuclear bomb that is set to explode. Action"
dataset_frequent_genres.iloc[33094,12]="Come Look at Me Oleg Yankovskiy Irina Kupchenko Yekaterina Vasilyeva Natalya Shchukina Mark Rudinshtejn Ivan Yankovskiy Mikhail Agranovich Before New Year Eve the windows of vast city are glowing from within by colored Christmas lights. It seems that behind each of these windows carefully stored favorite holiday. It seems that each of them are waiting for the New Year's wish fulfillment. By the silhouette of an elderly woman at one of the windows of the old house in Moscow, neighbors in the yard probably accustomed as something unchangeable. Sophia Ivanovna ten years does not get up from her chair, all day looking out the window, glues paper figurines and listening to Dickens, who reads aloud to her, her only daughter Tanya. The only heiress, Tanya, whom Sophia Ivanovna must to pass all the family jewels, seems resigned to the position of an old maid, and her whole life consists only of the care of the sick mother Comedy Drama Romance"
dataset_frequent_genres.iloc[36101,12]="For Y'ur Height Only Weng Weng Yehlen Catral Carmi Martin Anna Marie Gutierrez Beth Sandoval Eddie Nicart Mr. Giant has kidnapped the brilliant Dr. Van Kohler and is planning to use the Doctor's invention, the N-bomb, to hold the world hostage. The only one who can foil Mr. Giant's evil scheme is Agent 00, a 3-foot-tall filipino martial arts master, expert marksman, top-class romancer and all-around superspy. Can Agent 00 rescue Dr. Kohler before it's too late? Action Comedy"
dataset_frequent_genres.iloc[36670,12]="Header Jake Suffian Elliot V. Kotek Dick Mullaney Michael Philip Anthony Stacey Brooks Tara Brooks Jim Coope Bill Corry Andrew Cowen Amanda Czelinski Stephen DeCaires Kevin Dedes Lauren Devlin Ruth Dimino Morris Fazzi Jr. Archibald Flancranstin HEADER portrays the grueling psychological journey taken by ATF Agent Stewart Cummings. On the surface, Stewart struggles to solve a string of bizarre murders, but in secret, his life falls into a world of corruption that's impossible to escape. Deceit, rape, and murder spiral out of control triggering a hellish conclusion that defies description. Horror"
dataset_frequent_genres.iloc[47726,12]="Deewangee Ajay Devgn Akshaye Khanna Urmila Matondkar Farida Jalal Vijayendra Ghatge Seema Biswas Tiku Talsania Tanaaz Currim Irani Mohan Kapoor Nirmal Pandey Nishigandha Wad Rana Jung Bahadur Suresh Oberoi Sushovan Banerjee Suhasini Mulay Anees Bazmee Raj is one of city's top lawyer known never to loose a case. He his introduced to Sargam during the launch of her new album by Ashwin Mehta owner of a music company. Next day Ashwin is found murdered and Tarang is held for the murder, Sargam and Tarang are childhood friends and she approaches Raj to fight for his case. During this period Raj notices that Tarang suffers from mental illness of split personality and when he suffers the attack he is not Tarang but Ranjeet and doesn't remember things he does when he turns Ranjeet. Tarang had killed Ashwin as the previous night he saw him forcing on Sargam but doesn't remember killing him as he was Ranjeet that time. Raj wins the case in court by proving Tarang innocent as he is suffering from mental illness. Tarang is due to release and gives Raj a shock to that he had planned everything regarding split personality as he was caught red handed on the site of murder. Crime Drama Mystery"

# DELETE the movie: 6 Days to Air: The Making of South Park because its a short tv series.
dataset_frequent_genres=dataset_frequent_genres[dataset_frequent_genres['title']!="6 Days to Air: The Making of South Park"]
dataset_frequent_genres=dataset_frequent_genres.reset_index(drop=True)

In [43]:
"""Print the index of the dataframe. This is only for demonstration purposes."""
print(dataset_frequent_genres.index)

RangeIndex(start=0, stop=49122, step=1)


In [44]:
"""Serialize the dataset to local disk space for use."""
joblib.dump(dataset_frequent_genres,"dataset_part_2_cleaned_of_redundant_genres_18022021.pkl")

['dataset_part_2_cleaned_of_redundant_genres_18022021.pkl']

From this point it is advised to check Part 2.1-Connecting & Enriching dataset with movies.\
The next part will be used to enrich the dataset with more movies or add more content for each movie.\
The output dataset of part 2.1 will be used in part 3.1.

#### THIS IS THE END OF PART 2 - Data Transformation & Cleaning